### PR TITLE
[PM] Set market end when admin closes the market

### LINKED ISSUE #942: After using adminMoveMarketToClosed still need to wait for end
When using adminMoveMarketToClosed and then trying to report the outcome, extrisic fails with the message shown in the screenshot. The market has gracePeriod set to 0.

image

image

It would be perfect if the `range.end` in the storage is changed as well to the moment this extrinsic executed. We might need this for automated tests eventually, but for now it is fine if it stays the same so manual testing can be quickly done.

### PR DESCRIPTION
**Pull request by Chralt98**
Fixes #942

This is related to @TvrtkoM 's [comment on Discord](https://discord.com/channels/737780518313000960/817041223201587230/1062655318363996192).

> Really good point. The error is based on the actual market period end, which was put in by the market creator. Because you used adminMoveMarketToClosed this end was not edited to be at the time of the call to adminMoveMarketToClosed. To be able to report, you currently need to wait until the real market end time is reached.

> The question, when solving the problem is: Should we modify the MarketPeriod(range.end) to be at the time the market close was set by the admin or should we rather leave it unmodified and add another storage field.

I decided to modify the end, because I see no negative consequences at the moment.

This allows to [report the market directly](https://github.com/zeitgeistpm/zeitgeist/blob/e991d9edbc76db6f2ecc2780a7caf4a4702cff6a/zrml/prediction-markets/src/lib.rs#L1169-L1174) after closing it as an admin.

### CODE REVIEW

File: zrml/prediction-markets/src/lib.rs [hunk 1]
```diff
@@ -2345,6 +2346,25 @@ mod pallet {
...
+                market.period = match market.period {
+                    MarketPeriod::Block(ref range) => {
+                        let current_block = <frame_system::Pallet<T>>::block_number();
+                        MarketPeriod::Block(range.start..current_block)
+                    }
+                    MarketPeriod::Timestamp(ref range) => {
+                        let current_time_frame = Self::calculate_time_frame_of_moment(
+                            <zrml_market_commons::Pallet<T>>::now(),
+                        );
+                        MarketPeriod::Timestamp(range.start..current_time_frame)
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:2360]: The `Timestamp` variant of `MarketPeriod` actually holds the UNIX timestamp (in ms), not the time frame.

```suggestion
                        let now = <zrml_market_commons::Pallet<T>>::now();
                        MarketPeriod::Timestamp(range.start..now)
```

File: zrml/prediction-markets/src/tests.rs [hunk 2]
```diff
@@ -104,19 +104,25 @@ fn simple_create_scalar_market(
...
-            0..2,
+            now..end,
             ScoringRule::CPMM,
         );
+        run_blocks(3);
         let market_id = 0;
         assert_ok!(PredictionMarkets::admin_move_market_to_closed(Origin::signed(SUDO), market_id));
         let market = MarketCommons::market(&market_id).unwrap();
         assert_eq!(market.status, MarketStatus::Closed);
+        let new_end = now + 3;
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/tests.rs:123]: I think this should be (probably missing a cast here :wink:):

```suggestion
        let new_end = now + 3 * MILLISECS_PER_BLOCK;
```

**Review by maltekliemann — CHANGES_REQUESTED** [review]
I think this solution is fine. It messes a bit with the history of the market, but these changes can be traced back through an indexer. Plus we're only using these functions on the testnet.

**Comment by maltekliemann** [comment]
Actually, this solution is perfectly idiomatic. It just feels confusing because we don't use a cache to determine when we're allowed to report.

File: zrml/prediction-markets/src/tests.rs (hunk 2 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/tests.rs:123]: We are using `MarketPeriod::Block` here. If we had used `MarketPeriod::Timestamp` here I would totally agree.

**Comment by Chralt98** [comment]
>  we don't use a cache to determine when we're allowed to report.

Yeah, we currently only cache when the market closes and derive that the report does go after it. But because I don't see any negative consequences to this solution, I would go with it. Do you see any negative consequences?

File: zrml/prediction-markets/src/tests.rs (hunk 2 above)
Comment (maltekliemann) [reply zrml/prediction-markets/src/tests.rs:123]: Oh, my bad.

Comment (maltekliemann) [reply zrml/prediction-markets/src/tests.rs:123]: So there's no test for this when using timestamps? If not, could you add one? Seems important.

**Comment by maltekliemann** [comment]
> > we don't use a cache to determine when we're allowed to report.
> 
> Yeah, we currently only cache when the market closes and derive that the report does go after it. But because I don't see any negative consequences to this solution, I would go with it. Do you see any negative consequences?

No, I think this is absolutely the correct solution.

File: zrml/prediction-markets/src/tests.rs (hunk 2 above)
Comment (Chralt98) [reply zrml/prediction-markets/src/tests.rs:123]: ad3f275465481a1d7bd95990f883c5b550bac3d9

**Review by maltekliemann — APPROVED** [review]
There is one concern here, which is why these functions should absolutely not be used for anything but testing. Since the grace period is specified in terms of _duration_ rather than a timepoint, the grace period of a market that is closed with `admin_move_market_to_closed` will end prematurely. Not a problem, unless you're dealing with a real market.

https://github.com/zeitgeistpm/zeitgeist/blob/ad3f275465481a1d7bd95990f883c5b550bac3d9/zrml/prediction-markets/src/lib.rs#L1198-L1200

As for the question about possible side effects. The thing we need to look out for is that it doesn't break the `market_status_manager`. Closing the market forcefully clears the auto-close cached in `MarketIdsPerClose*`, and even if it didn't that wouldn't be a problem because `on_market_close` takes care not to mess with a market that is in an unexpected state.

**Review by sea212 — COMMENTED** [review]
PR LGTM.

In regards to https://github.com/zeitgeistpm/zeitgeist/pull/943#pullrequestreview-1269136107
> There is one concern here, which is why these functions should absolutely not be used for anything but testing. Since the grace period is specified in terms of duration rather than a timepoint, the grace period of a market that is closed with admin_move_market_to_closed will end prematurely. Not a problem, unless you're dealing with a real market.

I think we should handle every function in a pallet as if it was used in a production environment, unless it is clearly only enabled in specific environments such as `test`.
How should we differentiate between the different environments they are deployed in, for example if it is in BS or Zeitgeist? I think that what you Malte meant with "real" markets, the ones on the mainnet are real and the ones on BS are not, right? Do you think there is a way to adjust the grace period properly?

Besides that, those admin functions are not used on the mainnet, are they? In that case, would somebody please open an issue to filter them there?

**Comment by maltekliemann** [comment]
> I think we should handle every function in a pallet as if it was used in a production environment, unless it is clearly only enabled in specific environments such as `test`. How should we differentiate between the different environments they are deployed in, for example if it is in BS or Zeitgeist? I think that what you Malte meant with "real" markets, the ones on the mainnet are real and the ones on BS are not, right? Do you think there is a way to adjust the grace period properly?

Just to clarify, by "testing purposes" I mean using these function on Battery Station for playing with markets. Yes, we can easily adjust the grace period. The only question is if that's "correct". For example, if I want to test the grace period works correctly on Battery Station, it would be more natural if its length would remain unchanged (instead of its end)?

> Besides that, those admin functions are not used on the mainnet, are they? In that case, would somebody please open an issue to filter them there?

They've never been used on the mainnet and I would strongly suggest to filter them in the future.

**Review by sea212 — APPROVED** [review]

**Comment by sea212** [comment]
As a side note, when developing pallets always assume they are used by other projects. If there are some side effects during a function invocation, we'll have to document it. We should also mark when a function is not adequate for a production environment.

**Pull request merged**